File: 0437-path-sum-iii/0437-path-sum-iii.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} targetSum
 * @return {number}
 */
var pathSum = function(root, targetSum) {
    return countPathsWithSum('og', root, 0, targetSum, 0, {});
};

function countPathsWithSum(direction, node, currentSum, targetSum, pathCount, instancesOfSumValues) {
    if (!node) return 0;

    currentSum = currentSum + node.val;
    
    var newCount = 0;
    
    // console.log('instancesOfSumValues: ', instancesOfSumValues);
    const countOfTargetDelta = instancesOfSumValues[currentSum - targetSum] || 0;

    // console.log('moved in direction: ', direction);
    // console.log('current node val: ', node.val);
    // console.log('currentSum: ', currentSum);
    // console.log('targetSum: ', targetSum);
    // console.log('currentSum - targetSum: ', currentSum - targetSum);
    if (countOfTargetDelta > 0) {
        // console.log('target delta match - increasing count by: ', countOfTargetDelta);
        newCount += countOfTargetDelta;
    }

    
    if (currentSum === targetSum) {
        // console.log('currentSum = targetSum - increasing count by 1');
        newCount += 1;
    }
    
    // console.log();
    
    const countOfCurrentSum = instancesOfSumValues[currentSum] || 0;
    instancesOfSumValues[currentSum] = countOfCurrentSum + 1;
        
    var leftPathCount = countPathsWithSum('left', node.left, currentSum, targetSum, pathCount, {...instancesOfSumValues});
    var rightPathCount = countPathsWithSum('right', node.right, currentSum, targetSum, pathCount, {...instancesOfSumValues});
    
    // console.log('leftPathCount: ', leftPathCount);
    // console.log('leftPathCount: ', rightPathCount);
    
    newCount += leftPathCount + rightPathCount;

    return newCount;

}

File: 0437-path-sum-iii/0437-path-sum-iii.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {  
    public int pathSumFromNode(TreeNode node, long currentSum, int targetSum) {
        if (node == null) return 0;
        
        currentSum = currentSum + node.val;
        // System.out.println("targetSum: " + targetSum + ", currentSum: " + currentSum);
                
        int leftCount = pathSumFromNode(node.left, currentSum, targetSum);
        int rightCount = pathSumFromNode(node.right, currentSum, targetSum);
        
        int pathCount = leftCount + rightCount;
        
        if (currentSum == targetSum) pathCount += 1;
        
        return pathCount;
    }
    
    public int pathSum(TreeNode root, int targetSum) {
        if (root == null) return 0;
        
        int pathsFromNode = pathSumFromNode(root, 0, targetSum);
        
        int leftCount = pathSum(root.left, targetSum);
        int rightCount = pathSum(root.right, targetSum);
        
        return pathsFromNode + leftCount + rightCount;
    }
}